synthtool/gcp/templates/python_notebooks_testing_pipeline/.cloud-build/execute_notebook.py

```diff
@@ -0,0 +1,85 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import sys
+import nbformat
+from nbconvert.preprocessors import ExecutePreprocessor
+import os
+import errno
+from notebook_processors import RemoveNoExecuteCells, UpdateVariablesPreprocessor
+from typing import Dict
+import papermill as pm
+
+# This script is used to execute a notebook and write out the output notebook.
+# The replaces calling the nbconvert via command-line, which doesn't write the output notebook correctly when there are errors during execution.
+
+
+def execute_notebook(
+    notebook_file_path: str, output_file_folder: str, replacement_map: Dict[str, str]
+):
+    file_name = os.path.basename(os.path.normpath(notebook_file_path))
+
+    # Read notebook
+    with open(notebook_file_path) as f:
+        nb = nbformat.read(f, as_version=4)
+
+    has_error = False
+
+    # Execute notebook
+    try:
+        # Create preprocessors
+        remove_no_execute_cells_preprocessor = RemoveNoExecuteCells()
+        update_variables_preprocessor = UpdateVariablesPreprocessor(
+            replacement_map=replacement_map
+        )
+        ExecutePreprocessor(timeout=-1, kernel_name="python3")
+
+        # Use no-execute preprocessor
+        (nb, resources,) = remove_no_execute_cells_preprocessor.preprocess(nb)
+
+        (nb, resources) = update_variables_preprocessor.preprocess(nb, resources)
+
+        # Execute notebook
+        # out = execute_preprocessor.preprocess(nb, resources)
+        output_file_path = os.path.join(
+            output_file_folder, "failure" if has_error else "success", file_name
+        )
+
+        # Create directories if they don't exist
+        if not os.path.exists(os.path.dirname(output_file_path)):
+            try:
+                os.makedirs(os.path.dirname(output_file_path))
+            except OSError as exc:  # Guard against race condition
+                if exc.errno != errno.EEXIST:
+                    raise
+
+        print(f"Writing modified notebook to: {output_file_path}")
+        with open(output_file_path, mode="w", encoding="utf-8") as f:
+            nbformat.write(nb, f)
+
+        pm.execute_notebook(
+            input_path=output_file_path,
+            output_path=output_file_path,
+            progress_bar=True,
+            request_save_on_cell_execute=True,
+            log_output=True,
+            stdout_file=sys.stdout,
+            stderr_file=sys.stderr,
+        )
+
+    except Exception:
```

Comment:
I'm a little skittish about catching ALL exceptions, but at least we're raising them so I think I feel better about that. Do other @googleapis/python-samples-owners have opinions here?

Reply:
If an exception isn't caught, then Cloud Build will fail on that step, basically failing all the notebooks (even ones that would have otherwise passed.) 

Also, the cleanup and debug steps at the end of the Cloud Build pipeline will never run, thus the user wouldn't be able to debug the error.

Reply:
Actually, this whole file seems out of date, it should be matching this file: https://github.com/GoogleCloudPlatform/ai-platform-samples/blob/master/.cloud-build/ExecuteNotebook.py